### PR TITLE
List style feature

### LINKED ISSUE #7801: Implement editing part for List Styles feature
Part of https://github.com/ckeditor/ckeditor5/issues/1031.

### PR DESCRIPTION
**Pull request by pomek**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Feature (list): Introduced the list styles feature that allows customizing the list's marker of the list item elements. Closes #7801.

Feature (theme-lark): Creates styles for the `ListStylesUI` plugin (see #7803).

MINOR BREAKING CHANGE (ui): It is now possible to override existing components when [adding new ones](https://ckeditor.com/docs/ckeditor5/latest/api/module_ui_componentfactory-ComponentFactory.html#function-add) to the [component factory](https://ckeditor.com/docs/ckeditor5/latest/api/module_ui_componentfactory-ComponentFactory.html) (previously an error was thrown) (see #7803).

---

### Additional information

UI part of the feature was reviewed in https://github.com/ckeditor/ckeditor5/pull/7845. 
Docs: https://github.com/ckeditor/ckeditor5/pull/7881 (can be merged after this PR).

### CODE REVIEW

**Comment by pomek** [comment]
Keep in mind that I had some issues with the implementation and the feature could contain some bugs. See https://github.com/ckeditor/ckeditor5/issues/7879.

**Comment by pomek** [comment]
@ckeditor/qa-team, could you check how critical are issues described in the #7879 issue? I mean, are we able to merge the feature core and fix bugs later?

File: packages/ckeditor5-list/src/liststylesui.js
Comment (pomek) [inline packages/ckeditor5-list/src/liststylesui.js:220]: Shouldn't we wrap the code in the `model#change()` block in order to have single undo step?

**Comment by Reinmar** [comment]
One thing to keep in mind here is that the number of edge cases and weird cases and unclear cases will be significant. I'd heavily prioritize stability over correctness.

The other thing is that we're approaching the code freeze, so minor code refactorings should be split to a separate issue.

File: packages/ckeditor5-list/src/liststyles.js
Comment (pomek) [inline packages/ckeditor5-list/src/liststyles.js:15]:
```suggestion
 * The list styles feature.
```

**Comment by Mgsy** [comment]
## Steps to reproduce

1.  Select multiple cells in a table.
2.  Apply a list style.

## Current result

The regular list is applied.

## GIF

**Comment by Mgsy** [comment]
## Steps to reproduce

1.  Set the following data - `<ol><li>test 1</li></ol><ul><li>test 2</li></ul><ol><li>test 3</li></ol>`.
2.  Select all list items.
3.  Apply some list style.

## Current result

Only the first item receives styling.

## Expected result

All lists will be merged.

## GIF

File: packages/ckeditor5-list/src/liststyles.js
Comment (pomek) [inline packages/ckeditor5-list/src/liststyles.js:18]:
```suggestion
 * and the {@link module:list/liststylesui~ListStylesUI list styles UI feature}.
```

**Comment by Mgsy** [comment]
## Steps to reproduce

1.  Change a paragraph directly to a styled list.

## Current result

Two undo steps are created.

## Expected result

One undo step is created.

**Comment by Mgsy** [comment]
## Steps to reproduce

1.  Select some text in a list item.
2.  Add bold.

## Current result

The editor crashes.

## Error

```
liststylesediting.js:97 Uncaught TypeError: Cannot read property 'getStyle' of null
    at UpcastDispatcher.dispatcher.on.priority (liststylesediting.js:97)
    at UpcastDispatcher.fire (emittermixin.js:209)
    at UpcastDispatcher._convertItem (upcastdispatcher.js:249)
    at UpcastDispatcher.convert (upcastdispatcher.js:207)
    at Object.callback (datacontroller.js:414)
    at Model._runPendingChanges (model.js:830)
    at Model.change (model.js:175)
    at DataController.toModel (datacontroller.js:413)
    at MutationHandler._handleContainerChildrenMutations (injecttypingmutationshandling.js:109)
    at MutationHandler.handle (injecttypingmutationshandling.js:64)
```

**Comment by FilipTokarski** [comment]
### Steps: 
1. Add two styled list items 
2. Apply link on the second one
3. Select the second one and press `Bulleted list` on the toolbar

**Expected:** list changes to unordered

**Actual:** list changes to unordered only when you choose the specific style

01_list3

**Comment by Mgsy** [comment]
## Steps to reproduce

1.  Create a new list with one nested item ( `<ol><li>Test<ol><li>Test 2</li></ol></li></ol>` ).
2.  Select all items.
3.  Add styling.

## Current result

Only the root item is changed.

**Comment by FilipTokarski** [comment]
If you select a few indented list items, toggling betweend bulleted and numbered styles all of them. But if you want to add some specific styling - only the first item gets styled:  
01_list4

File: packages/ckeditor5-list/src/liststylescommand.js
Comment (jodator) [inline packages/ckeditor5-list/src/liststylescommand.js:162]: :+1:  for styling :)

File: packages/ckeditor5-list/src/liststylesediting.js
Comment (jodator) [inline packages/ckeditor5-list/src/liststylesediting.js:93]: I'd go with a name that reflects upcasting list item's style. The name `upcastListItem()` might suggest that it handles whole `<li>` upcast.

File: packages/ckeditor5-list/src/liststylesui.js
Comment (jodator) [reply packages/ckeditor5-list/src/liststylesui.js:220]: It might do the trick AFAIR. So for now might be OK.

The other thing that came to my mind might be that `ListSyyle` command _might_ also do the `parentCommand` stuff but as per PR requirements to not-block this PR a wrap will be quicker.

File: packages/ckeditor5-ui/src/componentfactory.js
Comment (jodator) [inline packages/ckeditor5-ui/src/componentfactory.js:92]: This is a change that probably would need to be discussed.

Comment (jodator) [reply packages/ckeditor5-ui/src/componentfactory.js:92]: I've scanned the #7803 but I didn't find why this change.

Comment (jodator) [reply packages/ckeditor5-ui/src/componentfactory.js:92]: OK so this is mostly to @oleq - I don't get from the issue why. The benefits are OK though.

@pomek - please ignore my comments here ;)

File: packages/ckeditor5-list/src/liststylescommand.js
Comment (jodator) [inline packages/ckeditor5-list/src/liststylescommand.js:57]: This might cause "single item changed on range selection" issues.

From top of my head:

1.  Get elements/blocks from selection (to would get all selected list items).
2.  Use `selection.getFirst/LastPostion()` to get backward/forward siblings.

**Review by jodator — CHANGES_REQUESTED** [review]
I've saw that there are some issues found so maybe checking if using `differ` instead of command's events might be relevant.

Abut naming - I've asked that question on the Slack channel as I'm not sure why the plural form and I might be lacking some conversation details.

If I commented something in the UI part - please ignore that. I've forgot instantly that this PR was partially reviewed. I would have some insights for retro though :D 

I'll do the jsdocs/tests re-read later on.

---

To sum up and to not block:

1.  The feature works, with minor bugs. Which, at least IMHO, we could pass.
2.  The naming might be the most important one before releasing (PITA to do, but should be fairly fast).
3.  crashes to be addressed/checked first.
4.  The event thing (if no time left) should be at least hidden under @Private to have it refactor later as I'm almost 100% sure that we could do the cleanup without it.

**Comment by jodator** [comment]
So I've also found similar issues to the above:



The 2-commands action behaves strangely because the first command (list type) changes all items to the other type and the second command (list style) applies style only to the top-most item.

So I think that my comment regarding how this command should work is valid (take elements from range or just the same ones as list type command does).

ps.: I've might found a bug in the original command as the one item does not change it's type :open\_mouth: .

**Comment by jodator** [comment]
> ps.: I've might found a bug in the original command as the one item does not change it's type open\_mouth .

It's some kind of downcast rendering glitch. I can't reproduce it on fresh content.

File: packages/ckeditor5-list/src/indentcommand.js
Comment (pomek) [reply packages/ckeditor5-list/src/indentcommand.js]: I marked the event as private and added `_` as prefix. (0143084)
But you're right. Post-fixer could be better solution here.

File: packages/ckeditor5-list/src/liststylescommand.js
Comment (pomek) [reply packages/ckeditor5-list/src/liststylescommand.js]: Thx :D I was wondering how to explain what I am doing in particular functions. I guess this "trick" explains everything.

**Comment by pomek** [comment]
So as for now, we have three main issues:

1. the command changes the most outdented list and ignores nested ones (assuming that selection has some)
2. sometimes render fails (refreshing the same scenario helps)
3. `_executeCleanup` pattern is not correct but it could be done as a followup after merging the PR.

> The naming might be the most important one before releasing (PITA to do, but should be fairly fast).

Now the feature is called `ListStyle`.

**Comment by jodator** [comment]
@pomek I've updated an unrelated test name as it was bugging me ;)

**Comment by jodator** [comment]
@Reinmar if @pomek doesn't need to add anything else here I think that this might go to the testing phase as well.

**Review by jodator — APPROVED** [review]

**Comment by pomek** [comment]
Let me add a few tests that you mentioned and I'll check whether `ListStyleCommand` has all use cases covered.

**Comment by pomek** [comment]
@jodator, I added missing stuff. It's ready for merge.

**Pull request merged**
